### ISSUE TITLE
Support maptiles based on GCJ02 (Chinese)

### ISSUE THREAD

**Issue by johnd0e** [IITC-CE/ingress-intel-total-conversion#64]
**Background info**:
* https://en.wikipedia.org/wiki/Restrictions_on_geographic_data_in_China#GCJ-02
*  https://web.archive.org/web/20140116215242/http://home.wangjianshuo.com/archives/20081109_all_maps_in_china_are_transformed.htm

**Summary**: some map tiles providers use GCJ-02 coordinates, thus map appears with some offset.

_Originally posted by @modos189 in https://github.com/_render_node/MDExOlB1bGxSZXF1ZXN0MjQ0Njc3OTI3/pull_requests/unread_timeline#issuecomment-454773564_

**Comment by johnd0e** [comment]
There are currently two attempts to fix the problem.
* ~PR #60 wip, _trying to change the coordinates of portal, link and field_ (seems not good way).~
* [fix-google-map-offsets-in-china](https://github.com/IITC-CE/ingress-intel-total-conversion/commits/fix-google-map-offsets-in-china) (~still nonfunctional~)

**Comment by johnd0e** [comment]
Some links from former discussion:
* https://github.com/htoooth/Leaflet.ChineseTmsProviders (offset problem)
* https://github.com/hobo0cn/leaflet-gcj02 (Perhaps this solution is for autonavi.com, but Google maps we connect in a completely different way)
* https://github.com/sureleo/leaflet-baidu

**Comment by johnd0e** [comment]
@modos189 could you explain the reason for this change: https://github.com/IITC-CE/ingress-intel-total-conversion/commit/6d3e28493cf5ba1ecc59b8d3432f99f8501b3b53?

Also I'm interesting what is the source of transforming algorithm.
I found similar:
* https://github.com/wandergis/coordtransform
* https://github.com/googollee/eviltransform/blob/master/javascript/transform.js
* https://github.com/Artoria2e5/PRCoords
* https://github.com/hiwanz/wgs2mars.js
* https://github.com/hujiulong/gcoord

And here are also different _'point-in-polygon'_ algorithm implementations: https://github.com/IITC-CE/ingress-intel-total-conversion/pull/65#issuecomment-455151352

**Comment by modos189** [comment]
> @modos189 could you explain the reason for this change

Previously, the fix coordinates included most of India, all of South and North Korea, the Philippines, Vietnam, Mongolia, Thailand, and a number of other countries:
http://programmerz.ru/images/76604/BG9LI.jpg

Code in https://github.com/IITC-CE/ingress-intel-total-conversion/commit/6d3e28493cf5ba1ecc59b8d3432f99f8501b3b53 more accurately indicates the borders of China, so as not to affect other countries. I took it here: https://github.com/iitc-project/ingress-intel-total-conversion/pull/1188

**Comment by modos189** [comment]
[Here](https://github.com/IITC-CE/ingress-intel-total-conversion/commit/64060c4b78564c5d0d7d40f0603b48560cbd01ac) I managed to partially solve the problem.

**BIG plus 1:** Do not change the coordinates of portals
**BIG plus 2:** No changes to other plugins are made

**Minus:** 
With a zoom of 16, to the right sometimes not loaded tiles.
With a zoom of 18, this gets worse and the right side of the screen is already without tiles.

The fact is that due to the correction of coordinates the map moves and goes beyond the boundaries for which the tiles are loaded.

By the way, now the correction of coordinates applies to all maps (except satellite and hybrid)

**Comment by johnd0e** [comment]
Is it possible to provide this solution to other basemap-plugins?

> Code in 6d3e284 more accurately indicates the borders of China

So there is complex polygon now. Doesn't it affect the performance? Or maybe better make 2-steps check?
**edit**: I see `isInGoogle` is that 1st step that I mean.

>  the right side of the screen is already without tiles.

Do you have ideas how to fix it?

**Comment by johnd0e** [comment]
> Also I'm interesting what is the source of transforming algorithm.

Well, I see that the source is [mentioned](https://github.com/iitc-project/ingress-intel-total-conversion/pull/1188/files#diff-a669276f457868338ca6e6691c52556fL67) in that PR:
```
// The algorithm of transforming WGS-84 to GCJ-02 comes from:
// https://on4wp7.codeplex.com/SourceControl/changeset/view/21483#353936
```

So I wonder why these lines were removed?

P.S.
Just for reference: here are related author's blog's entries:
* [Conversion algorithm from Earth Coordinate System (WGS-84) to Mars Coordinate System (GCJ-02)](https://blog.csdn.net/coolypf/article/details/8686588)
* [Conversion algorithm between Mars coordinate system (GCJ-02) and Baidu coordinate system (BD-09)](https://blog.csdn.net/coolypf/article/details/8569813)

**Comment by modos189** [comment]
> Is it possible to provide this solution to other basemap-plugins?

Yep, that's what I want to do.

> So there is complex polygon now. Doesn't it affect the performance?

Visually, I do not notice the difference.

> **edit**: I see `isInGoogle` is that 1st step that I mean.

I don't know what that is. And it's not used anywhere.
 
> > the right side of the screen is already without tiles.
>
> Do you have ideas how to fix it?

I fixed this in all maps except Google. Now I do not know how to fix

> So I wonder why these lines were removed?

Mistakenly deleted. Now returned.

**Comment by johnd0e** [comment]
I found error in the code (mistype). Also I have concerns about some style's elements.
So I will prepare PR.

**Comment by modos189** [comment]
Since there are both maps that require coordinate correction and those that do not require it, I suggest pass an optional parameter when initializing the map.
Example:

```
            var mapLayer = new L.TileLayer(baseUrl + address, {
                attribution: 'map tiles by autonavi.com, plugin written by GMOogway.',
                subdomains: '1234',
                layer: layer,
                type: '',
                minZoom: minZoom,
                maxNativeZoom: maxZoom,
                maxZoom: 20,

                needFixChinaOffset: true
            });

            layerChooser.addBaseLayer(mapLayer, '高德 ' + name);
```

**Comment by modos189** [comment]
I found a solution to fix google maps

**Comment by johnd0e** [comment]
> I suggest pass an optional parameter when initializing the map.

I would be perfect solution.

But I still have concerns about leaflet monkey-patching in `fix-googlemap-china-offset` script.
I wonder if it could be implemented via leaflet plugin instead.

**Comment by johnd0e** [comment]
We need to make here a list of known maptiles providers, in order to check:
* which tiles need correction, and which don't.
* does area of _Hong Kong and Macau_ is distorted or not.

Even more, we need 2 lists: first for providers included with iitc, and second for all others.

**Comment by johnd0e** [comment]
Included:
- [x] Google.com (except `satellite` and `hybrid`)
- [ ] CartoDB
- [ ] gmaps-gray
- [ ] Bing
- [ ] Kartverket
- [ ] Here WeGo (no public access to Chinese maptiles)
- [ ] OpenCycleMap
- [ ] ~~OpenStreetMap~~
- [ ] Stamen
- [ ] Yandex

**Comment by johnd0e** [comment]
Chinese:
- [x] Google.cn
- [x] Baidu
- [x] AutoNavi
- [x] NavInfo?
- [x] TianDiTu
- [x] GaoDe
- [x] Geoq
- [x] Soso
- [x] Alibaba Cloud
- [x] Tencent
- [ ] ...

Others:
- [ ] MapQuest
- [ ] Mapbox
- [ ] Esri
- [ ] Thunderforest
- [ ] ...

**Comment by magicbug2013** [comment]
@johnd0e 
Hong Kong and Macau: WGS84
google map(except satellite and hybrid): gcj02
baidu map: bd09
gaode map and other china maps: gcj02
all the other maps are correct.

**Comment by johnd0e** [comment]
> Hong Kong and Macau: WGS84

There could be difference depending on map. In current code (#72) HK positions are also processed for Baidu map.
So we really need to check HK for every Chinese map, to make sure that it's Baidu-only.

> gaode map and other china maps: gcj02

We also need to check all their variations (incl. satellite/hybrid).
Do you have sample plugins for other china maps?

> all the other maps are correct.

According Wikipedia [Here](https://wego.here.com/) are also in GSJ02. Could you check it?

**Comment by magicbug2013** [comment]
> > Hong Kong and Macau: WGS84
> 
> There could be difference depending on map. In current code (#72) HK positions are also processed for Baidu map.
> So we really need to check HK for every Chinese map, to make sure that it's Baidu-only.
> 
> > gaode map and other china maps: gcj02
> 
> Do you have sample plugins for other china maps?
> 
> > all the other maps are correct.
> 
> According Wikipedia [Here](https://wego.here.com/) are also in GSJ02. Could you check it?

Too complicated? I think we can use gedo-map + fix-china-map-gcj02-offset in China. Otherwise, we can use Google map directly without fix-china-map-gcj02-offset. We don't use baidu-map (bd09), and now baidu-map can't be displayed in intel.

baidu map in intel:
![image](https://user-images.githubusercontent.com/5212258/51436280-2c5df180-1cc5-11e9-88cd-6282be2374d2.png)

**Comment by johnd0e** [comment]
> Otherwise, we can use Google map directly without fix-china-map-gcj02-offset

We fix-china-map-offset for Google, don't we?

> baidu map in intel:

How have you got this screen?

**Comment by magicbug2013** [comment]
> We fix-china-map-offset for Google, don't we?

yes, but if I have gaode map in china, why use google map? google map update is not in timely, In China, gaode map is enough.

> How have you got this screen?

I used the same method as gaode map, but it didn't display correctly.

**Comment by johnd0e** [comment]
> gaode map is enough

Ok. And what is difference between 'gaode' and 'autonavi'?

> I used the same method as gaode map, but it didn't display correctly.

May be this can help: https://github.com/muyao1987/leaflet-tileLayer-baidu

**Comment by magicbug2013** [comment]
>Ok. And what is difference between 'gaode' and 'autonavi'?

gaode=autonavi ^^

>May be this can help: https://github.com/muyao1987/leaflet-tileLayer-baidu

ok

**Comment by johnd0e** [comment]
Ok. So could you check Nokia/Here map (https://github.com/iitc-project/ingress-intel-total-conversion/pull/1262)?

**Comment by johnd0e** [comment]
Interesting, there is official converting api:
* https://lbs.amap.com/api/javascript-api/reference/lnglat-to-address/#m_AMap.convertFrom
* http://lbsyun.baidu.com/index.php?title=webapi/guide/changeposition

**Comment by modos189** [comment]
May be to use 2 parameters in which to pass a list of layers that need correction?
`needFixChinaGCJ02Offset: ['roads']`
or
`needFixChinaBD09Offset: ['roads']` // only for baidu

**Comment by johnd0e** [comment]
> May be to use 2 parameters in which to pass a list of layers that need correction?

As I stated here https://github.com/IITC-CE/ingress-intel-total-conversion/pull/72#issuecomment-455914716, we need to test some real maps before we could make proper decisions.

1. It's possible that we could leave `hybrid` and `satellite` excludes internal (for google), and not expose any option for this.
2. As well, I have not seen Baidu map yet, so I am not sure if we ever need some special handling for it.
It's possible that we just remove all baidu-related stuff.

**Comment by johnd0e** [comment]
Baidu-related links:
* http://tiven.wang/articles/use-baidu-map-provider-in-leafletjs/
* http://jsfiddle.net/roxeteer/L4yef9pr/3/
* https://github.com/maptalks/raster-collection

**Comment by johnd0e** [comment]
Fixed by #68
